src/driver/op.rs

```diff
@@ -147,46 +174,92 @@ where
                 *lifecycle = Lifecycle::Ignored(Box::new(self.data.take()));
             }
             Lifecycle::Completed(..) => {
                 inner.ops.remove(self.index);
             }
+            Lifecycle::CompletionList(indices) => {
+                // Deallocate list entries, recording if the more CQE's are expected
+                let more = {
+                    let mut list = indices.into_list(completions);
+                    io_uring::cqueue::more(list.peek_end().unwrap().flags)
+                    // Dropping list deallocates the list entries
+                };
+                if more {
+                    // If more are expected, we have to keep the op around
+                    *lifecycle = Lifecycle::Ignored(Box::new(self.data.take()));
+                } else {
+                    inner.ops.remove(self.index);
+                }
+            }
             Lifecycle::Ignored(..) => unreachable!(),
         }
     }
 }
 
 impl Lifecycle {
-    pub(super) fn complete(&mut self, cqe: CqeResult) -> bool {
+    pub(super) fn complete(&mut self, completions: &mut Slab<Completion>, cqe: CqeResult) -> bool {
         use std::mem;
 
         match mem::replace(self, Lifecycle::Submitted) {
-            Lifecycle::Submitted => {
-                *self = Lifecycle::Completed(cqe);
+            x @ Lifecycle::Submitted | x @ Lifecycle::Waiting(..) => {
+                if io_uring::cqueue::more(cqe.flags) {
+                    let mut list = SlabListIndices::new().into_list(completions);
+                    list.push(cqe);
+                    *self = Lifecycle::CompletionList(list.into_indices());
+                } else {
+                    *self = Lifecycle::Completed(cqe);
+                }
+                if let Lifecycle::Waiting(waker) = x {
+                    // wake is called whenever we have more work.
+                    // Also possible would be to defer calling until cqe with !more flag set
+                    waker.wake();
+                }
                 false
             }
-            Lifecycle::Waiting(waker) => {
-                *self = Lifecycle::Completed(cqe);
-                waker.wake();
+            lifecycle @ Lifecycle::Ignored(..) => {
+                // We must check if any more CQEs are expected before dropping
+                if io_uring::cqueue::more(cqe.flags) {
+                    *self = lifecycle;
+                    false
+                } else {
+                    true
+                }
+            }
+
+            Lifecycle::Completed(..) => {
+                // Completions with more flag set go straight onto the slab,
+                // and are handled in Lifecycle::CompletionList.
+                // To construct Lifecycle::Completed, a CQE without MORE was received
+                // we shouldn't be receiving another.
+                unreachable!("invalid operation state")
+            }
+            Lifecycle::CompletionList(indices) => {
```

Comment:
This case also causes non-More cqe to be pushed onto the list. We only expect one, but a second one doesn't get noticed here and doesn't cause an unreachable panic like the case above does. That difference bothers me enough that I think some comment is warranted here, about both kinds of cqe being pushed at least.